flatnav/util/VisitedNodesHandler.h

```diff
@@ -0,0 +1,175 @@
+  }
+
+  // move constructor
+  VisitedNodesHandler(VisitedNodesHandler &&other) noexcept
+      : _table_size(other._table_size), _mark(other._mark),
+        _table(other._table) {
+    other._table = nullptr;
+    other._table_size = 0;
+    other._mark = 0;
+  }
+
+  // copy assignment
+  VisitedNodesHandler &operator=(const VisitedNodesHandler &other) {
+    if (this != &other) {
+      delete[] _table;
+      _table_size = other._table_size;
+      _mark = other._mark;
+      _table = new uint32_t[_table_size];
+      std::memcpy(_table, other._table, _table_size * sizeof(uint32_t));
+    }
+    return *this;
+  }
+
+  // move assignment
+  VisitedNodesHandler &operator=(VisitedNodesHandler &&other) noexcept {
+    _table_size = other._table_size;
+    _mark = other._mark;
+    _table = other._table;
+    other._table = NULL;
+    other._table_size = 0;
+    other._mark = 0;
+    return *this;
+  }
+};
+
+/**
+ *
+ * @brief Manages a pool of VisitedNodesHandler objects in a thread-safe manner.
+ *
+ * This class is designed to efficiently provide and manage a pool of
+ * VisitedNodesHandler instances for concurrent use in multi-threaded
+ * environments. It ensures that each handler can be used by only one thread at
+ * a time without the risk of concurrent access and modification.
+ *
+ * The class preallocates a specified number of VisitedNodesHandler objects to
+ * eliminate the overhead of dynamic allocation during runtime. It uses a mutex
+ * to synchronize access to the handler pool, ensuring that only one thread can
+ * modify the pool at any given time. This mechanism provides both thread safety
+ * and improved performance by reusing handler objects instead of continuously
+ * creating and destroying them.
+ *
+ * When a thread requires a VisitedNodesHandler, it can call
+ * `pollAvailableHandler()` to retrieve an available handler from the pool. If
+ * the pool is empty, the function will dynamically allocate a new handler to
+ * ensure that the requesting thread can proceed with its task. Once the thread
+ * has finished using the handler, it should return it to the pool by calling
+ * `pushHandler()`.
+ *
+ * @note The class assumes that all threads will properly return the handlers to
+ * the pool after use. Failing to return a handler will deplete the pool and
+ * lead to dynamic allocation, negating the performance benefits.
+ *
+ * Usage example:
+ * @code
+ * ThreadSafeVisitedNodesHandler handler_pool(10, 1000);
+ * VisitedNodesHandler* handler = handler_pool.pollAvailableHandler();
+ * // Use the handler in a thread...
+ * handler_pool.pushHandler(handler);
+ * @endcode
+ *
+ * @param initial_pool_size The number of handler objects to initially create
+ * and store in the pool.
+ * @param num_elements The size of each VisitedNodesHandler, which typically
+ * corresponds to the number of nodes or elements that each handler is expected
+ * to manage.
+ */
+class ThreadSafeVisitedNodesHandler {
```

Comment:
Would it be correct to call these classes "VisitedNodeSet" and an "AtomicVisitedNodeSet"? or maybe Pool*

Other options: VisitedList / VisitedListPool, VisitedSet / VisitedSetPool, etc.

Reply:
I like VisitedSet and VisistedSetPool.